### DISCUSSION TITLE
datasets.create for VSAM

### DISCUSSION THREAD

**Discussion by gngrossi** [IBM/zoau-samples#58] [Q&A]
Having an issue creating VSAM datasets. Attached is a file from the REPL.
Commands issued:

datasets.create(name=N1, type="KSDS", primary_space="1M", secondary_space="1M", block_size=4096, record_format="FB", record_length=80, key_length=4, key_offset=0)

datasets.create(name=N1, type="ESDS", primary_space="1M", secondary_space="1M", block_size=4096, record_format="FB", record_length=80)

datasets.create(name=N1, type="RRDS", primary_space="1M", secondary_space="1M", block_size=4096, record_format="FB", record_length=80)

# ESDS, RRDS, LDS get defined but with an IndexError
datasets.create(name=N1, type="ESDS")

# does not get defined with the 
[ZOAU_datasets.create.txt](https://github.com/IBM/zoau-samples/files/10007480/ZOAU_datasets.create.txt)
IndexError
datasets.create(name=N1, type="KSDS")

thanks

**Comment by EnsonoStevenPerva** [comment]
So, to add some color to this problem. I'm having the same issue with the provided code. I modified the code to look like the following:
```
from zoautil_py import datasets
val = datasets._create(name="SYS1.MY.KSDS", type="KSDS", primary_space="1M", secondary_space="1M", block_size=4096, record_format="FB", record_length=80, key_length=4, key_offset=0, debug=True)
print(val.to_dict())
```

On my system, this resulted in:
```
{'rc': 4, 'stdout_response': '', 'stderr_response': 'BGYSC1402E Dynalloc Failure. Error Code: 0x9700, Information Code: 0x00\n', 'command': 'dtouch -d -tKSDS -s1M -e1M -B4096 -rFB -l80 -k4:0 SYS1.MY.KSDS'}
```
I then issued the raw command in the prompt adding the `a` and `v` options:
`dtouch -dav -tKSDS -s1M -e1M -B4096 -rFB -l80 -k4:0 SYS1.MY.KSDS`
which left me with this:
```
dtouch -dav -tKSDS -s1M -e1M -B4096 -rFB -l80 -k4:0 SYS1.MY.KSDS
 __ddname:
 __status:    ???
 __alcunit:   TRK
 __dsorg:     ???
 __dsntype:   EXTPREF
 __storclass:
 __mgntclass:
 __dataclass: {0}
 __primary:   17
 __secondary: 17
 __recfm:     FB
 __lrecl:     80
 __blksize:   4096
 __recorg:    KSDS
 __rls:       ??
 __keyoffset: 0
 __keylength: 4
BGYSC1402E Dynalloc Failure. Error Code: 0x9700, Information Code: 0x00
```

Hope this helps someone troubleshoot.

**Reply by gngrossi** [reply]
It seems like the VSAM dataset support isn't quite there yet,, hopefully it will be soon.
Here's an example of what we use for IDCAMS statements for ESDS, RRDS, KSDS and LDS. Options SHR and CISIZE would need to be supported.

DEFINE CLUSTER                       -   
       (NAME(@02858.T.SDT.DEMO.ESDS) -   
        RECORDSIZE(60 60)            -   
        CYLINDERS (5 1)              -   
        SHR (2 3)                    -   
        NONINDEXED )                 -   
  DATA                               -   
       (NAME(@02858.T.SDT.DEMO.ESDS.D) - 
       CONTROLINTERVALSIZE (4096))       

DEFINE CLUSTER                       -   
       (NAME(@02858.T.SDT.DEMO.RRDS) -   
        RECORDSIZE(60 60)            -   
        CYLINDERS (5 1)              -   
        SHR (2 3)                    -   
        NUMBERED )                   -   
  DATA                               -   
       (NAME(@02858.T.SDT.DEMO.RRDS.D) - 
       CONTROLINTERVALSIZE (4096))       

DEFINE CLUSTER                     -   
       (NAME(CICS.T.SDT.DEMO.KSDS) -   
        KEYS (4 0)                 -   
        RECORDSIZE(60 60)          -   
        CYLINDERS (5 1)            -   
        SHR (2 3)                  -   
        INDEXED )                  -   
  DATA                             -   
       (NAME(CICS.T.SDT.DEMO.KSDS.D) - 
       CONTROLINTERVALSIZE (4096))   - 
  INDEX                              - 
       (NAME(CICS.T.SDT.DEMO.KSDS.I) - 
       CONTROLINTERVALSIZE (2048))     

DEFINE CLUSTER (NAME(CICS.T.SYS9.LDSS.ZZZ#01) -
                CYL(2,1) -
                SHAREOPTIONS(1,3) -  
                LINEAR)

thanks

**Reply by 2020jyc** [reply]
any update on this topic?  I saw only create dataset includes VSAM with ZOAU 1.2.x.    Is there any update on reading VSAM dataset?
